Mathlib/Data/Bitvec/Defs.lean

```diff
@@ -1,339 +1,177 @@
-
-instance (n : ℕ) : Preorder (Bitvec n) :=
-  Preorder.lift Bitvec.toNat
-
-/-- convert `fin` to `Bitvec` -/
-def ofFin {n : ℕ} (i : Fin <| 2 ^ n) : Bitvec n :=
-  Bitvec.ofNat _ i.val
-#align bitvec.of_fin Bitvec.ofFin
-
-/-- convert `Bitvec` to `fin` -/
-def toFin {n : ℕ} (i : Bitvec n) : Fin (2 ^ n) :=
-  i.toNat
-#align bitvec.to_fin Bitvec.toFin
-
-
-/-- Return the integer encoded by the input bitvector -/
-protected def toInt : ∀ {n : Nat}, Bitvec n → Int
-  | 0, _ => 0
-  | succ _, v =>
-    cond (head v) (Int.negSucc <| Bitvec.toNat <| Bitvec.not <| tail v)
-      (Int.ofNat <| Bitvec.toNat <| tail v)
-#align bitvec.to_int Bitvec.toInt
-
-end Conversion
-
-/-! ### Miscellaneous instances -/
-
-
-private def repr {n : Nat} : Bitvec n → String
-  | ⟨bs, _⟩ => "0b" ++ (bs.map fun b : Bool => if b then '1' else '0').asString
-
-instance (n : Nat) : Repr (Bitvec n) where
-  reprPrec (b : Bitvec n) _ := Std.Format.text (repr b)
-
-end Bitvec
-
-instance {n} {x y : Bitvec n} : Decidable (Bitvec.Ult x y) :=
-  decEq _ _
+/-!
+## Arithmetic
+We inherit `Fin` implementations.
+-/
 
-instance {n} {x y : Bitvec n} : Decidable (Bitvec.Ugt x y) :=
-  decEq _ _
+/-- Less than for two bitvectors. -/
+protected def lt (x y : BitVec w) : Bool :=
+  x.val < y.val
 
-instance {n} : HShiftLeft (Bitvec n) Nat (Bitvec n) := ⟨Bitvec.shl⟩
+/-- Less than or equal to for two bitvectors. -/
+protected def le (x y : BitVec w) : Bool :=
+  x.val ≤ y.val
 
-instance {n} : HShiftRight (Bitvec n) Nat (Bitvec n) := ⟨Bitvec.ushr⟩
+instance : Ring (BitVec w) := inferInstanceAs (Ring (Fin _))
+instance : Mod (BitVec w) := inferInstanceAs (Mod (Fin _))
+instance : Div (BitVec w) := inferInstanceAs (Div (Fin _))
+instance : LT (BitVec w)  := ⟨fun x y => BitVec.lt x y⟩
+instance : LE (BitVec w)  := ⟨fun x y => BitVec.le x y⟩
 
-instance {n} : ShiftLeft (Bitvec n) := ⟨fun x y => x <<< y.toNat⟩
+/-!
+## Bitwise operations
+-/
 
-instance {n} : ShiftRight (Bitvec n) := ⟨fun x y => x >>> y.toNat⟩
+/-- The complement of a bitvector. -/
+protected def complement (x : BitVec w) : BitVec w :=
+  0 - (x + .ofNat w 1)
+
+/-- The and of two bitvectors. -/
+protected def and (x y : BitVec w) : BitVec w :=
+  ⟨x.val &&& y.val, by simp [HAnd.hAnd, AndOp.and, Nat.land, Nat.bitwise_lt]⟩
+
+/-- The or of two bitvectors. -/
+protected def or (x y : BitVec w) : BitVec w :=
+  ⟨x.val ||| y.val, by simp [HOr.hOr, OrOp.or, Nat.lor, Nat.bitwise_lt]⟩
+
+/-- The xor of two bitvectors. -/
+protected def xor (x y : BitVec w) : BitVec w :=
+  ⟨x.val ^^^ y.val, by simp [HXor.hXor, Xor.xor, Nat.xor, Nat.bitwise_lt]⟩
+
+/-- The left-shift of a bitvector by some amount. -/
+protected def shiftLeft (x : BitVec w) (n : Nat) : BitVec w :=
+  .ofNat w (x.val <<< n)
+
+/-- The right-shift of a bitvector by some amount. -/
+protected def shiftRight (x : BitVec w) (n : Nat) : BitVec w :=
+  ⟨x.val >>> n, by
+      simp only [Nat.shiftRight_eq_div_pow]
+      exact lt_of_le_of_lt (Nat.div_le_self' _ _) (x.isLt) ⟩
```

Comment:
Either you should remove these definitions (and inline their implementation into the `instance` below), or you should add a simp lemma for each that replaces it with the corresponding notation.